signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,314 @@
+                // we don't need to do anything so we return early.
+                if self.is_seed_address(&addr) {
+                    tracing::debug!(peer_id = %e.peer_id, %addr, "disconnected from seed");
+                } else {
+                    return;
+                }
+
+                // Update our connected seeds map.
+                if let Some(set) = self.connected_seeds.get_mut(&e.peer_id) {
+                    if set.remove(&addr) {
+                        tracing::trace!("removed connected seed: {:?}", e.peer_id);
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+    }
+
+    #[tracing::instrument(skip_all)]
+    fn poll(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p::swarm::ToSwarm<Self::ToSwarm, libp2p::swarm::THandlerInEvent<Self>>>
+    {
+        // If we have any pending events, we return them immediately.
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        // If we have any connections, we consider ourselves bootstrapped.
+        // Determine if there was a state change or not and return the correct
+        // poll result.
+        if self.connection_count >= 1 {
+            return match self.is_bootstrapped {
+                true => Poll::Pending,
+                false => {
+                    // We've just bootstrapped, so we generate an event to notify the
+                    // swarm.
+                    tracing::debug!(connection_count = %self.connection_count, "bootstrapping complete");
+                    self.is_bootstrapped = true;
+                    Poll::Ready(ToSwarm::GenerateEvent(BootstrapEvent::Complete))
+                }
+            };
+        }
+
+        // If we're here then we're not connected to any peers. If our current
+        // state is bootstrapped, we need to re-bootstrap.
+        if self.is_bootstrapped {
+            tracing::debug!(
+                "state is bootstrapped but not connected to any peers; re-bootstrapping is needed"
+            );
+            self.is_bootstrapped = false;
+            self.pending_events
+                .push_back(ToSwarm::GenerateEvent(BootstrapEvent::Needed));
+        }
+
+        // If we've attempted to bootstrap recently, we wait until the interval
+        // has passed.
+        if let Some(last_bootstrap) = self.last_attempted_at {
+            if last_bootstrap.elapsed() < self.config.bootstrap_interval {
+                return Poll::Pending;
+            }
+        }
+
+        // Queue the bootstrap started event.
+        tracing::debug!(addresses = ?self.config.seed_addresses, "initiating network bootstrapping from seed addresses");
```

Comment:
nit: since we already log this event in the libp2p event loop, isn't this a duplicated log?

Reply:
I'd like to keep this one for the time being -- my plan is that we move more and more of the event loop into behaviors, thus removing that logging.

Reply:
I moved all of the logging into the behavior and removed it from the event loop.